core/include/detray/utils/ranges/subrange.hpp

```diff
@@ -103,24 +74,113 @@ class subrange : public detray::ranges::view_interface<subrange<range_t>> {
     iterator_t m_end;
 };
 
+}  // namespace detail
+
+/// @brief interface type to construct a @c subrange_view with CTAD
+template <detray::ranges::range range_t, typename index_range_t = dindex_range>
+struct subrange : public detail::subrange_view<range_t> {
+
+    using base_type = detray::ranges::detail::subrange_view<range_t>;
+
+    using iterator_t = typename detray::ranges::iterator_t<range_t>;
+    using difference_t = typename detray::ranges::range_difference_t<range_t>;
+
+    /// Default constructor
+    constexpr subrange() = default;
+
+    /// Construct from a @param range and an index range @param pos.
+    template <concepts::index index_t>
+    DETRAY_HOST_DEVICE constexpr subrange(index_t start, index_t end)
+        : m_pos{start, end} {}
+
+    /// Construct from an @param start and @param end iterator pair.
+    DETRAY_HOST_DEVICE constexpr subrange(iterator_t start, iterator_t end)
+        : base_type{start, end} {}
+
+    /// Construct from a @param range.
+    template <detray::ranges::range deduced_range_t>
+    DETRAY_HOST_DEVICE constexpr explicit subrange(deduced_range_t &&range)
+        : base_type{detray::ranges::begin(std::forward<deduced_range_t>(range)),
+                    detray::ranges::end(std::forward<deduced_range_t>(range))} {
+    }
+
+    /// Construct from a @param range and starting position @param pos. Used
+    /// as an overload when only a single position is needed.
+    template <detray::ranges::range deduced_range_t, concepts::index index_t>
+    requires std::is_convertible_v<
+        index_t, detray::ranges::range_difference_t<deduced_range_t>>
+        DETRAY_HOST_DEVICE constexpr subrange(deduced_range_t &&range,
+                                              index_t pos)
+        : base_type{
+              detray::ranges::next(
+                  detray::ranges::begin(std::forward<deduced_range_t>(range)),
+                  static_cast<difference_t>(pos)),
+              detray::ranges::next(
+                  detray::ranges::begin(std::forward<deduced_range_t>(range)),
+                  static_cast<difference_t>(pos) + difference_t{1})} {}
+
+    /// Construct from a @param range and an index range @param pos.
+    template <detray::ranges::range deduced_range_t,
+              concepts::interval deduced_index_range_t>
+    DETRAY_HOST_DEVICE constexpr subrange(deduced_range_t &&range,
+                                          deduced_index_range_t &&pos)
+        : base_type{get_itr<0>(std::forward<deduced_range_t>(range),
+                               std::forward<deduced_index_range_t>(pos)),
+                    get_itr<1>(std::forward<deduced_range_t>(range),
+                               std::forward<deduced_index_range_t>(pos))} {}
```

Comment:
Note that this is dangerous; if `deduced_range_t` or `deduced_index_range_t` is an rvalue reference then this is a double move.

Reply:
Oops, thanks.